lib/db/open.go

```diff
@@ -0,0 +1,96 @@
+package db
+
+import (
+	"context"
+	"database/sql"
+	"math/rand"
+	"time"
+
+	_ "github.com/lib/pq"
+)
+
+type dbOpts struct {
+	maxConnectionLifetime       time.Duration
+	maxConnectionLifetimeJitter time.Duration
+	pingInterval                time.Duration
+	pingJitter                  time.Duration
+}
+
+type PGOpenOption func(*dbOpts)
+
+// WithMaxConnectionLifetime configures the maximum connection
+// lifetime. The current default value is set based on the known
+// pg-gateway and a2-backend timeouts. Consider carefully before
+// changing. A zero duration disables the connection lifetime.
+func WithMaxConnectionLifetime(d time.Duration) PGOpenOption {
+	return func(opts *dbOpts) {
+		opts.maxConnectionLifetime = d
+	}
+}
+
+// WithMaxConnectionLifetimeJitter adds a random duration in the range
+// of [0, d) to the configured MaxConnectionLifetime if it is
+// non-zero.
+func WithMaxConnectionLifetimeJitter(d time.Duration) PGOpenOption {
+	return func(opts *dbOpts) {
+		opts.maxConnectionLifetimeJitter = d
+	}
+}
+
+// WithPingInterval configures a background gorountine that
+// will ping the database every interval. A zero duration disables the
+// pinger. Default: 20s
+func WithPingInterval(d time.Duration) PGOpenOption {
+	return func(opts *dbOpts) {
+		opts.pingInterval = d
+	}
+}
+
+// WithoutPinger disables background pinging.
+func WithoutPinger() PGOpenOption { return WithPingInterval(0) }
+
+// WithPingJitter adds a random duration in the range of [o,
+// d) to the PeriodicPingInterval. Default: 20s
+func WithPingJitter(d time.Duration) PGOpenOption {
+	return func(opts *dbOpts) {
+		opts.pingJitter = d
+	}
+}
+
+func PGOpen(dataSourceName string, opts ...PGOpenOption) (*sql.DB, error) {
+	return PGOpenContext(context.Background(), dataSourceName, opts...)
+}
+
+func PGOpenContext(ctx context.Context, dataSourceName string, userOpts ...PGOpenOption) (*sql.DB, error) {
+	opts := &dbOpts{
+		maxConnectionLifetime:       15 * time.Minute,
+		maxConnectionLifetimeJitter: 15 * time.Minute,
+		pingInterval:                20 * time.Second,
```

Comment:
This enables the pinger by default.  Given that this only happens after 12 hours of inactivity, I'm not sure it is worth it.  But I suppose people using a2-cluster will hit this more often since their timeout is lower and this might help there.  I'm kinda loath to set maxConnectionLifetime to less than a minute...

Reply:
IMO enabling the pinger gets us closer to the optimal outcome and the price is low. It should help keep our tests and such stable as well which is a pretty solid benefit.